lib/src/test/java/org/kpipe/consumer/OffsetManagerTest.java

```diff
@@ -0,0 +1,739 @@
+
+    // First commit
+    final var firstCommand = performCommitAndCaptureCommand(commandQueue);
+
+    // Verify the first commit completed successfully
+    offsetManager.notifyCommitComplete(firstCommand.getCommitId(), true);
+
+    // Verify partition state after first commit
+    assertEventually(
+      () ->
+        expectedOffsetAfterFirstCommit == (long) offsetManager.getPartitionState(partition).get("nextOffsetToCommit"),
+      Duration.ofSeconds(2),
+      "Offset after first commit should be " + expectedOffsetAfterFirstCommit
+    );
+
+    // Track and process record with gap offset
+    final var largeGapRecord = new ConsumerRecord<>(TOPIC, 0, gapOffset, "key", "value");
+    offsetManager.trackOffset(largeGapRecord);
+    offsetManager.markOffsetProcessed(largeGapRecord);
+
+    // Second commit
+    final var secondCommand = performCommitAndCaptureCommand(commandQueue);
+
+    // Complete the second commit
+    offsetManager.notifyCommitComplete(secondCommand.getCommitId(), true);
+
+    // Verify partition state after second commit
+    assertEventually(
+      () ->
+        expectedOffsetAfterSecondCommit == (long) offsetManager.getPartitionState(partition).get("nextOffsetToCommit"),
+      Duration.ofSeconds(2),
+      "Offset after second commit should be " + expectedOffsetAfterSecondCommit
+    );
+  }
+
+  private ConsumerCommand performCommitAndCaptureCommand(final BlockingQueue<ConsumerCommand> queue) throws Exception {
+    final var commandLatch = new CountDownLatch(1);
+    final var commandRef = new AtomicReference<ConsumerCommand>();
+
+    // Start a command listener thread
+    Thread
+      .ofPlatform()
+      .daemon()
+      .start(() -> {
+        try {
+          final var cmd = queue.poll(10, TimeUnit.SECONDS);
+          if (cmd != null) {
+            commandRef.set(cmd);
+            commandLatch.countDown();
+          }
+        } catch (final InterruptedException e) {
+          // Ignore interruption
+        }
+      });
+
+    // Start the commit operation
+    CompletableFuture.runAsync(() -> {
+      try {
+        offsetManager.commitSyncAndWait(10);
+      } catch (final InterruptedException e) {
+        // Ignore interruption
+      }
+    });
+
+    // Wait for command
+    final var commandReceived = commandLatch.await(10, TimeUnit.SECONDS);
+    assertTrue(commandReceived, "Command should be produced within timeout");
+
+    final var command = commandRef.get();
+    assertNotNull(command, "Commit command should be generated");
+
+    return command;
+  }
+
+  private void assertEventually(final Supplier<Boolean> condition, final Duration timeout, final String message)
+    throws Exception {
+    final var deadline = System.currentTimeMillis() + timeout.toMillis();
+
+    while (System.currentTimeMillis() < deadline) {
+      if (condition.get()) {
+        return;
+      }
+      Thread.sleep(50);
+    }
+
+    fail(message);
+  }
+
+  static Stream<Arguments> gapSizeScenarios() {
```

Comment:
@rkolesnev

Reply:
the backwards gap test is invalid - due to Kafka guarantees - the other ones seem fine.

Reply:
fixed